staging/src/k8s.io/apiserver/pkg/admission/plugin/resourcequota/controller.go

```diff
@@ -131,15 +131,28 @@ func NewQuotaEvaluator(quotaAccessor QuotaAccessor, ignoredResources map[schema.
 		stopCh:  stopCh,
 		config:  config,
 	}
+
+	// The queue underneath is starting a goroutine for metrics
+	// exportint that is only stopped on calling ShutDown.
+	// Given that QuotaEvaluator is created for each layer of apiserver
+	// and often not started for some of those (e.g. aggregated apiserver)
+	// we explicitly shut it down on stopCh signal even if it wasn't
+	// effectively started.
+	go evaluator.shutdownOnStop()
```

Comment:
does it means we start only once but we shutdown multiple times?

Reply:
Sorry - maybe I wasn't clear.
What I meant is that:

- we create a separate QuotaEvaluator for every layer (separate for kube-apiserver, separate for agregated, separate for apiextensions)
- the creation of it actually creates a queue which starts a goroutine that is updating metrics (this is created in line 125 so once per quota evaluator)
- the newly added shutdownOnStop is also created here so called exactly once (per QuotaEvaluator)
- the start() is called *at most* once - line 606

Reply:
and why is not valid to keep it as it was?

Reply:
Not sure I understand the question.

As I wrote above - when I create QuotaEvaluator, this is underneath starting a gorotuine (created by `workqueue.NewNamed`).
That goroutine is shut down only when run() (renamed to start()) method is called.

run() is called on the first `Evaluate` - if that is not called - the goroutine will be leaked.

Reply:
ahhhhhhhhhhhhhhhhhhhhhhhh ... sorry friday and lack coffeeeeeee